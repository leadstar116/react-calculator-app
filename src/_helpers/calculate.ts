import Big from 'big.js'
import isNumber from "./isNumber"
import { ThunkDispatch } from 'redux-thunk'
import { Action } from 'redux';
import { calculateReset, calculateUpdate } from '../_actions/calculate.actions';
import { CalculatorState } from '../_constants/calculate.interface'
import { alertFailure } from '../_actions/alert.actions';

type MyRootState = {};
type MyExtraArg = undefined;
type MyThunkDispatch = ThunkDispatch<MyRootState, MyExtraArg, Action>;

/**
 * Params: a button name and a calculator state object
 * Action: Run calculation by calling API endpoint.
 *
 * Calculate state object contains:
 *   total:String      the running total
 *   next:String       the next number to be operated on with the total
 *   operation:String  +, -, etc.
 */
const apiUrl = 'https://nestjs-calculator-backend-api.herokuapp.com/calculate'
const calculate = (state: CalculatorState, buttonName: string) => async (dispatch: MyThunkDispatch) => {
  if (buttonName === "AC") {
    return dispatch(calculateReset())
  }

  if (isNumber(buttonName)) {
    if (buttonName === "0" && state.next === "0") {
      return;
    }
    // If there is an operation, update next
    if (state.operation) {
      if (state.next) {
        return dispatch(calculateUpdate({
          ...state,
          next: state.next + buttonName,
        }))
      }
      return dispatch(calculateUpdate({
        ...state,
        next: buttonName,
      }))
    }
    // If there is no operation, update next and clear the value
    if (state.next) {
      const next = state.next === "0" ? buttonName : state.next + buttonName;
      return dispatch(calculateUpdate({
        ...state,
        next: next,
        total: "",
      }))
    }
    return dispatch(calculateUpdate({
      ...state,
      next: buttonName,
      total: "",
    }))
  }

  if (buttonName === "%") {
    if (state.operation && state.next) {
      try {
        let response = await fetch(
          apiUrl,
          {
              method: 'POST',
              headers: { 'Content-type': 'application/json' },
              body: JSON.stringify(state)
          }
        )
        const result = await response.json()
        return dispatch(calculateUpdate({
          ...state,
          total: Big(result)
            .div(Big("100"))
            .toString(),
          next: "",
          operation: "",
        }))
      } catch(e) {
        return dispatch(alertFailure(e))
      }
    }
    if (state.next) {
      return dispatch(calculateUpdate({
        ...state,
        next: Big(state.next)
          .div(Big("100"))
          .toString(),
      }))
    }
    return;
  }

  if (buttonName === ".") {
    if (state.next) {
      // ignore a . if the next number already has one
      if (state.next.includes(".")) {
        return;
      }
      return dispatch(calculateUpdate({
        ...state,
        next: state.next + "."
      }))
    }
    return dispatch(calculateUpdate({
      ...state,
      next: "0."
    }))
  }

  if (buttonName === "=") {
    if (state.next && state.operation) {
      try {
        let response = await fetch(
          apiUrl,
          {
              method: 'POST',
              headers: { 'Content-type': 'application/json' },
              body: JSON.stringify(state)
          }
        )
        const result = await response.json()
        return dispatch(calculateUpdate({
          ...state,
          total: result,
          next: "",
          operation: "",
        }))
      } catch(e) {
        return dispatch(alertFailure(e))
      }
    } else {
      // '=' with no operation, nothing to do
      return;
    }
  }

  if (buttonName === "+/-") {
    if (state.next) {
      return dispatch(calculateUpdate({
        ...state,
        next: (-1 * parseFloat(state.next)).toString()
      }))
    }
    if (state.total) {
      return dispatch(calculateUpdate({
        ...state,
        total: (-1 * parseFloat(state.total)).toString()
      }))
    }
    return;
  }

  // Button must be an operation

  // When the user presses an operation button without having entered
  // a number first, do nothing.
  // if (!state.next && !state.total) {
  //   return {};
  // }

  // User pressed an operation button and there is an existing operation
  if (state.operation) {
    try {
      let response = await fetch(
        apiUrl,
        {
            method: 'POST',
            headers: { 'Content-type': 'application/json' },
            body: JSON.stringify(state)
        }
      )
      const result = await response.json()
      return dispatch(calculateUpdate({
        ...state,
        total: result,
        next: "",
        operation: buttonName,
      }))
    } catch(e) {
      return dispatch(alertFailure(e))
    }
  }

  // no operation yet, but the user typed one

  // The user hasn't typed a number yet, just save the operation
  if (!state.next) {
    return dispatch(calculateUpdate({
      ...state,
      operation: buttonName
    }))
  }

  // save the operation and shift 'next' into 'total'
  return dispatch(calculateUpdate({
    total: state.next,
    next: "",
    operation: buttonName,
  }))
}

export default calculate