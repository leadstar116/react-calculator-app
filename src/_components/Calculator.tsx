import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { initialState } from '../_constants/state.interface'
import calculate from '../_helpers/calculate'
import ButtonPanel from './ButtonPanel'
import Display from './Display'

const Calculator = () => {
  const state = useSelector((state:initialState) => state.calculateReducer)
  const dispatch = useDispatch()

  const handleClick = (buttonName: string) => {
    dispatch(calculate(state, buttonName))
  };

  return (
    <div className="calculator">
      <h2 className="text-center">Calculator</h2>
      <Display value={ state.next || state.total || "0"} />
      <ButtonPanel clickHandler={handleClick} />
    </div>
  );
}

export default Calculator