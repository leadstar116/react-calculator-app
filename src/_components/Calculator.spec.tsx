import configureStore from 'redux-mock-store';
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Calculator from './Calculator'
import { CalculatorState } from '../_constants/calculate.interface';

describe('Calculator Component', () => {
  const calculatorState = {
    next: "3",
    total: "2",
    operation: "+"
  } as CalculatorState;

  const mockStore = configureStore();
  let store = mockStore({
    calculateReducer: calculatorState,
  })

  store.dispatch = jest.fn();

  let component = renderer.create(
    <Provider store={store}>
      <Calculator />
    </Provider>
  );

  it('should render with given props', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
})