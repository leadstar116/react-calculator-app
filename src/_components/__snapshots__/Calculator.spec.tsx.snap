// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Component should render with given props 1`] = `
<div
  className="calculator"
>
  <h2
    className="text-center"
  >
    Calculator
  </h2>
  <div
    className="calculator-display"
  >
    <div>
      3
    </div>
  </div>
  <div
    className="calculator-button-panel"
  >
    <div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          AC
        </button>
      </div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          +/-
        </button>
      </div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          %
        </button>
      </div>
      <div
        className="calculator-button orange"
      >
        <button
          onClick={[Function]}
        >
          รท
        </button>
      </div>
    </div>
    <div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          7
        </button>
      </div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          8
        </button>
      </div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          9
        </button>
      </div>
      <div
        className="calculator-button orange"
      >
        <button
          onClick={[Function]}
        >
          x
        </button>
      </div>
    </div>
    <div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          4
        </button>
      </div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          5
        </button>
      </div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          6
        </button>
      </div>
      <div
        className="calculator-button orange"
      >
        <button
          onClick={[Function]}
        >
          -
        </button>
      </div>
    </div>
    <div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          1
        </button>
      </div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          2
        </button>
      </div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          3
        </button>
      </div>
      <div
        className="calculator-button orange"
      >
        <button
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
    <div>
      <div
        className="calculator-button wide"
      >
        <button
          onClick={[Function]}
        >
          0
        </button>
      </div>
      <div
        className="calculator-button"
      >
        <button
          onClick={[Function]}
        >
          .
        </button>
      </div>
      <div
        className="calculator-button orange"
      >
        <button
          onClick={[Function]}
        >
          =
        </button>
      </div>
    </div>
  </div>
</div>
`;
